File: DSA/DSA_450/Python/05_linked_list/004_Remove_Loop.py
"""
Problem: Remove Loop in Linked List
URL: https://practice.geeksforgeeks.org/problems/remove-loop-in-linked-list/1

Problem Statement:
Remove the loop from a linked list if it exists.

Sample Input/Output:
Input: 1->2->3->4->5->2 (loop at node 2)
Output: 1->2->3->4->5->NULL
Explanation: Loop is removed from the linked list.
"""

class ListNode:
    def __init__(self, x):
        self.data = x
        self.next = None

def Create_List(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for i in range(1, len(arr)):
        curr.next = ListNode(arr[i])
        curr = curr.next
    return head

def Create_Loop(head, pos):
    if pos < 0:
        return
    loop_node = None
    curr = head
    index = 0
    
    while curr.next:
        if index == pos:
            loop_node = curr
        curr = curr.next
        index += 1
    
    if loop_node:
        curr.next = loop_node

def List_To_Array(head, max_nodes=10):
    result = []
    count = 0
    while head and count < max_nodes:
        result.append(head.data)
        head = head.next
        count += 1
    return result

def Print_List(head, max_nodes=10):
    arr = List_To_Array(head, max_nodes)
    if arr:
        print("->".join(map(str, arr)), end="")
        if head and max_nodes == len(arr):
            print("->...", end="")
        print("->NULL")
    else:
        print("NULL")

class Solution:
    def Remove_Loop_Hashing(self, head):
        """
        Hashing approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        visited = set()
        prev = None
        curr = head
        
        while curr:
            if curr in visited:
                prev.next = None
                return
            visited.add(curr)
            prev = curr
            curr = curr.next
    
    def Remove_Loop_Floyd_Detect_Remove(self, head):
        """
        Floyd's detect and remove with counting
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if not head or not head.next:
            return
        
        slow = head
        fast = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                break
        
        if slow != fast:
            return
        
        loop_length = 1
        temp = slow.next
        while temp != slow:
            loop_length += 1
            temp = temp.next
        
        ptr1 = head
        ptr2 = head
        
        for i in range(loop_length):
            ptr2 = ptr2.next
        
        while ptr1 != ptr2:
            ptr1 = ptr1.next
            ptr2 = ptr2.next
        
        while ptr2.next != ptr1:
            ptr2 = ptr2.next
        
        ptr2.next = None
    
    def Remove_Loop_Floyd_Optimized(self, head):
        """
        Floyd's optimized without counting
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if not head or not head.next:
            return
        
        slow = head
        fast = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                break
        
        if slow != fast:
            return
        
        if slow == head:
            while fast.next != head:
                fast = fast.next
            fast.next = None
            return
        
        slow = head
        while slow.next != fast.next:
            slow = slow.next
            fast = fast.next
        
        fast.next = None

def Test_Remove_Loop():
    solution = Solution()
    
    arr1 = [1, 2, 3, 4, 5]
    head1 = Create_List(arr1)
    Create_Loop(head1, 1)
    print("Before removal (with loop): ", end="")
    Print_List(head1)
    solution.Remove_Loop_Hashing(head1)
    print("After removal (Hashing): ", end="")
    Print_List(head1)
    
    arr2 = [1, 2, 3, 4, 5]
    head2 = Create_List(arr2)
    Create_Loop(head2, 0)
    print("\nBefore removal (loop at head): ", end="")
    Print_List(head2)
    solution.Remove_Loop_Floyd_Detect_Remove(head2)
    print("After removal (Floyd with counting): ", end="")
    Print_List(head2)
    
    arr3 = [1, 2, 3, 4, 5, 6]
    head3 = Create_List(arr3)
    Create_Loop(head3, 2)
    print("\nBefore removal: ", end="")
    Print_List(head3)
    solution.Remove_Loop_Floyd_Optimized(head3)
    print("After removal (Floyd optimized): ", end="")
    Print_List(head3)

if __name__ == "__main__":
    Test_Remove_Loop()

File: DSA/DSA_450/05_linked_list/004_Remove_Loop.cpp
/*
Problem: Remove Loop in Linked List
URL: https://practice.geeksforgeeks.org/problems/remove-loop-in-linked-list/1

Problem Statement:
Remove the loop from a linked list if it exists.

Sample Input/Output:
Input: 1->2->3->4->5->2 (loop at node 2)
Output: 1->2->3->4->5->NULL
Explanation: Loop is removed from the linked list.
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int data;
    ListNode* next;
    ListNode(int x) : data(x), next(NULL) {}
};

ListNode* Create_List(vector<int>& arr) {
    if (arr.empty()) return NULL;
    ListNode* head = new ListNode(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < arr.size(); i++) {
        curr->next = new ListNode(arr[i]);
        curr = curr->next;
    }
    return head;
}

void Create_Loop(ListNode* head, int pos) {
    if (pos < 0) return;
    ListNode* loopNode = NULL;
    ListNode* curr = head;
    int index = 0;
    
    while (curr->next) {
        if (index == pos) {
            loopNode = curr;
        }
        curr = curr->next;
        index++;
    }
    
    if (loopNode) {
        curr->next = loopNode;
    }
}

void Print_List(ListNode* head, int maxNodes = 10) {
    int count = 0;
    while (head && count < maxNodes) {
        cout << head->data;
        if (head->next && count < maxNodes - 1) cout << "->";
        head = head->next;
        count++;
    }
    if (head) cout << "->...";
    cout << "->NULL" << endl;
}

class Solution {
public:
    void Remove_Loop_Hashing(ListNode* head) {
        /*
        Hashing approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        */
        unordered_set<ListNode*> visited;
        ListNode* prev = NULL;
        ListNode* curr = head;
        
        while (curr) {
            if (visited.find(curr) != visited.end()) {
                prev->next = NULL;
                return;
            }
            visited.insert(curr);
            prev = curr;
            curr = curr->next;
        }
    }
    
    void Remove_Loop_Floyd_Detect_Remove(ListNode* head) {
        /*
        Floyd's detect and remove with counting
        Time Complexity: O(n)
        Space Complexity: O(1)
        */
        if (!head || !head->next) return;
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            
            if (slow == fast) {
                break;
            }
        }
        
        if (slow != fast) return;
        
        int loopLength = 1;
        ListNode* temp = slow->next;
        while (temp != slow) {
            loopLength++;
            temp = temp->next;
        }
        
        ListNode* ptr1 = head;
        ListNode* ptr2 = head;
        
        for (int i = 0; i < loopLength; i++) {
            ptr2 = ptr2->next;
        }
        
        while (ptr1 != ptr2) {
            ptr1 = ptr1->next;
            ptr2 = ptr2->next;
        }
        
        while (ptr2->next != ptr1) {
            ptr2 = ptr2->next;
        }
        
        ptr2->next = NULL;
    }
    
    void Remove_Loop_Floyd_Optimized(ListNode* head) {
        /*
        Floyd's optimized without counting
        Time Complexity: O(n)
        Space Complexity: O(1)
        */
        if (!head || !head->next) return;
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            
            if (slow == fast) {
                break;
            }
        }
        
        if (slow != fast) return;
        
        if (slow == head) {
            while (fast->next != head) {
                fast = fast->next;
            }
            fast->next = NULL;
            return;
        }
        
        slow = head;
        while (slow->next != fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
        
        fast->next = NULL;
    }
};

void Test_Remove_Loop() {
    Solution solution;
    
    vector<int> arr1 = {1, 2, 3, 4, 5};
    ListNode* head1 = Create_List(arr1);
    Create_Loop(head1, 1);
    cout << "Before removal (with loop): ";
    Print_List(head1);
    solution.Remove_Loop_Hashing(head1);
    cout << "After removal (Hashing): ";
    Print_List(head1);
    
    vector<int> arr2 = {1, 2, 3, 4, 5};
    ListNode* head2 = Create_List(arr2);
    Create_Loop(head2, 0);
    cout << "\nBefore removal (loop at head): ";
    Print_List(head2);
    solution.Remove_Loop_Floyd_Detect_Remove(head2);
    cout << "After removal (Floyd with counting): ";
    Print_List(head2);
    
    vector<int> arr3 = {1, 2, 3, 4, 5, 6};
    ListNode* head3 = Create_List(arr3);
    Create_Loop(head3, 2);
    cout << "\nBefore removal: ";
    Print_List(head3);
    solution.Remove_Loop_Floyd_Optimized(head3);
    cout << "After removal (Floyd optimized): ";
    Print_List(head3);
}

int main() {
    Test_Remove_Loop();
    return 0;
}
